src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4117,220 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
+//   they are now a jump to the next instruction, this can happen when the following
+//   basic block emitted no instructions
+//
+// Returns:
+//   true if any jumps have been removed, otherwise returns false
+//
+// Assumptions:
+//    the jump list must be ordered by increasing igNum+insNo
+//
+bool emitter::emitRemoveJumpToNextInst()
+{
+#ifdef TARGET_XARCH
+    JITDUMP("*************** In emitRemoveJumpToNextInst()\n");
+#ifdef DEBUG
+    if (EMIT_INSTLIST_VERBOSE)
+    {
+        JITDUMP("\nInstruction group list before unconditional jump to next instruction removal:\n\n");
+        emitDispIGlist(true);
+    }
+    if (EMITVERBOSE)
+    {
+        emitDispJumpList();
+    }
+#endif
+
+    int            totalRemovedCount      = 0;
+    UNATIVE_OFFSET totalRemovedSize       = 0;
+    instrDescJmp*  jmp                    = emitJumpList;
+    instrDescJmp*  previousJmp            = nullptr;
+    insGroup*      previousTruncatedGroup = nullptr;
+#if DEBUG
+    UNATIVE_OFFSET previousJumpIgNum  = (UNATIVE_OFFSET)-1;
+    unsigned int   previousJumpInsNum = -1;
+#endif
+
+    while (jmp)
+    {
+        // if the jump is unconditional then it is a candidate
+        if (jmp->idInsFmt() == IF_LABEL && emitIsUncondJump(jmp) && jmp->idjIsJmpAlways)
+        {
+            insGroup* jmpGroup = jmp->idjIG;
+#if DEBUG
+            assert((jmpGroup->igFlags & IGF_HAS_ALIGN) == 0);
+            assert(jmpGroup->igNum > previousJumpIgNum || previousJumpIgNum == (UNATIVE_OFFSET)-1 ||
+                   (jmpGroup->igNum == previousJumpIgNum && jmp->idDebugOnlyInfo()->idNum > previousJumpInsNum));
+            previousJumpIgNum  = jmpGroup->igNum;
+            previousJumpInsNum = jmp->idDebugOnlyInfo()->idNum;
+#endif
+            // target group is not bound yet so use the cookie to fetch it
+            insGroup* targetGroup = (insGroup*)emitCodeGetCookie(jmp->idAddr()->iiaBBlabel);
+
+            if (targetGroup != nullptr && jmpGroup->igNext == targetGroup)
+            {
+                unsigned instructionCount = jmpGroup->igInsCnt;
+
+                if (instructionCount > 1)
+                {
+                    instrDesc* instructionDescriptor = nullptr;
+
+                    {
+                        BYTE* dataPtr = jmpGroup->igData;
+                        while (instructionCount > 0)
+                        {
+                            instructionDescriptor = (instrDesc*)dataPtr;
+                            dataPtr += emitSizeOfInsDsc(instructionDescriptor);
+                            instructionCount -= 1;
+                        };
+                        dataPtr = nullptr;
+                    }
+
+                    assert(instructionDescriptor != nullptr);
+                    // instructionDescriptor is now the last instruction in the group
+
+                    if (instructionDescriptor->idIns() == jmp->idIns() && jmp == instructionDescriptor)
```

Comment:
I don't know whether it could happen. Now I've tested it I can see that it doesn't happen in the tests but that doesn't mean it can't happen in the real world. I prefer to be cautions/defensive rather than assume coverage is perfect. I've changed this to an assert as requested because in theory it can't be harmful.